From 3cb76def571cc7f9ab35c33ebe1c3431778f46df Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Fri, 29 Aug 2014 21:18:41 -0400
Subject: [net] svcrpc: handle some gssproxy encoding errors

Message-id: <1409347125-11495-3-git-send-email-bfields@redhat.com>
Patchwork-id: 89915
O-Subject: [Patch RHEL7 2/6] svcrpc: handle some gssproxy encoding errors
Bugzilla: 1117914
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

Reported-by: Andi Kleen <andi@firstfloor.org>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
(cherry picked from commit b26ec9b11b309acd9f6bb15fcc9bb396091384e8)
Bugzilla 1117914

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/auth_gss/gss_rpc_xdr.c b/net/sunrpc/auth_gss/gss_rpc_xdr.c
index f0f78c5..1ec19f6 100644
--- a/net/sunrpc/auth_gss/gss_rpc_xdr.c
+++ b/net/sunrpc/auth_gss/gss_rpc_xdr.c
@@ -559,6 +559,8 @@ static int gssx_enc_cred(struct xdr_stream *xdr,
 
  /* cred->elements */
  err = dummy_enc_credel_array(xdr, &cred->elements);
+ if (err)
+  return err;
 
  /* cred->cred_handle_reference */
  err = gssx_enc_buffer(xdr, &cred->cred_handle_reference);
@@ -740,22 +742,20 @@ void gssx_enc_accept_sec_context(struct rpc_rqst *req,
   goto done;
 
  /* arg->context_handle */
- if (arg->context_handle) {
+ if (arg->context_handle)
   err = gssx_enc_ctx(xdr, arg->context_handle);
-  if (err)
-   goto done;
- } else {
+ else
   err = gssx_enc_bool(xdr, 0);
- }
+ if (err)
+  goto done;
 
  /* arg->cred_handle */
- if (arg->cred_handle) {
+ if (arg->cred_handle)
   err = gssx_enc_cred(xdr, arg->cred_handle);
-  if (err)
-   goto done;
- } else {
+ else
   err = gssx_enc_bool(xdr, 0);
- }
+ if (err)
+  goto done;
 
  /* arg->input_token */
  err = gssx_enc_in_token(xdr, &arg->input_token);
@@ -763,13 +763,12 @@ void gssx_enc_accept_sec_context(struct rpc_rqst *req,
   goto done;
 
  /* arg->input_cb */
- if (arg->input_cb) {
+ if (arg->input_cb)
   err = gssx_enc_cb(xdr, arg->input_cb);
-  if (err)
-   goto done;
- } else {
+ else
   err = gssx_enc_bool(xdr, 0);
- }
+ if (err)
+  goto done;
 
  err = gssx_enc_bool(xdr, arg->ret_deleg_cred);
  if (err)
-- 
1.7.1